codec/dagcbor/unmarshal.go

```diff
@@ -49,9 +50,23 @@ func (cfg DecodeOptions) Decode(na datamodel.NodeAssembler, r io.Reader) error {
 		return na2.DecodeDagCbor(r)
 	}
 	// Okay, generic builder path.
-	return Unmarshal(na, cbor.NewDecoder(cbor.DecodeOptions{
+	err := Unmarshal(na, cbor.NewDecoder(cbor.DecodeOptions{
 		CoerceUndefToNull: true,
 	}, r), cfg)
+
+	if err != nil {
+		return err
+	}
+
+	var buf [1]byte
+	count, err := r.Read(buf[:])
+	if count != 0 {
```

Comment:
From the io.Reader docs:

> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0. Callers should treat a return of 0 and nil as indicating that nothing happened; in particular it does not indicate EOF. 

So you could end up with Read calls returning the following:

```
0, nil
0, nil
1, nil
0, io.EOF
```

If you want to properly skip those "noop" reads, use `io.ReadFull`:

> The error is EOF only if no bytes were read. If an EOF happens after reading some but not all the bytes, ReadFull returns ErrUnexpectedEOF. On return, n == len(buf) if and only if err == nil.

Reply:
k, can you check the latest commit and comment on that approach?

Reply:
Done, approved the entire thing. Will let you merge as you please, e.g. squashing.